src/music/win32_m.cpp

```diff
@@ -15,170 +15,361 @@
-	UINT_PTR devid;
-	char start_song[MAX_PATH];
+	UINT time_period;
+	HMIDIOUT midi_out;
+	UINT timer_id;
+	CRITICAL_SECTION lock;
+
+	bool playing;        ///< flag indicating that playback is active
+	bool do_start;       ///< flag for starting playback of next_file at next opportunity
+	bool do_stop;        ///< flag for stopping playback at next opportunity
+	byte current_volume; ///< current effective volume setting
+	byte new_volume;     ///< volume setting to change to
+
+	MidiFile current_file;           ///< file currently being played from
+	PlaybackSegment current_segment; ///< segment info for current playback
+	DWORD playback_start_time;       ///< timestamp current file started playing back
+	size_t current_block;            ///< next block index to send
+	MidiFile next_file;              ///< upcoming file to play
+	PlaybackSegment next_segment;    ///< segment info for upcoming file
+
+	byte channel_volumes[16];
 } _midi;
 
 static FMusicDriver_Win32 iFMusicDriver_Win32;
 
-void MusicDriver_Win32::PlaySong(const char *filename)
+
+static byte ScaleVolume(byte original, byte scale)
 {
-	assert(filename != NULL);
-	strecpy(_midi.start_song, filename, lastof(_midi.start_song));
-	_midi.playing = true;
-	_midi.stop_song = false;
-	SetEvent(_midi.wait_obj);
+	return original * scale / 127;
 }
 
-void MusicDriver_Win32::StopSong()
+
+void CALLBACK MidiOutProc(HMIDIOUT hmo, UINT wMsg, DWORD_PTR dwInstance, DWORD_PTR dwParam1, DWORD_PTR dwParam2)
 {
-	if (_midi.playing) {
-		_midi.stop_song = true;
-		_midi.start_song[0] = '\0';
-		SetEvent(_midi.wait_obj);
+	if (wMsg == MOM_DONE) {
+		MIDIHDR *hdr = (LPMIDIHDR)dwParam1;
+		midiOutUnprepareHeader(hmo, hdr, sizeof(*hdr));
+		free(hdr);
 	}
 }
 
-bool MusicDriver_Win32::IsSongPlaying()
+static void TransmitChannelMsg(byte status, byte p1, byte p2 = 0)
 {
-	return _midi.playing;
+	midiOutShortMsg(_midi.midi_out, status | (p1 << 8) | (p2 << 16));
 }
 
-void MusicDriver_Win32::SetVolume(byte vol)
+static void TransmitSysex(byte *&msg_start, size_t &remaining)
 {
-	_midi.new_vol = vol;
-	SetEvent(_midi.wait_obj);
+	/* find end of message */
+	byte *msg_end = msg_start;
+	while (*msg_end != 0xF7) msg_end++;
+	msg_end++; /* also include sysex end byte */
+
+	/* prepare header */
+	MIDIHDR *hdr = CallocT<MIDIHDR>(1);
+	if (midiOutPrepareHeader(_midi.midi_out, hdr, sizeof(*hdr)) == MMSYSERR_NOERROR) {
+		/* transmit - just point directly into the data buffer */
+		hdr->lpData = (LPSTR)msg_start;
+		hdr->dwBufferLength = msg_end - msg_start;
+		hdr->dwBytesRecorded = hdr->dwBufferLength;
+		midiOutLongMsg(_midi.midi_out, hdr, sizeof(*hdr));
+	} else {
+		free(hdr);
+	}
+
+	/* update position in buffer */
+	remaining -= msg_end - msg_start;
+	msg_start = msg_end;
 }
 
-static MCIERROR CDECL MidiSendCommand(const TCHAR *cmd, ...)
+void CALLBACK TimerCallback(UINT uTimerID, UINT, DWORD_PTR dwUser, DWORD_PTR, DWORD_PTR)
 {
-	va_list va;
-	TCHAR buf[512];
+	if (TryEnterCriticalSection(&_midi.lock)) {
```

Comment:
What's the effect of not being able to acquire the lock?

Reply:
Good point, I'll write some comments there. The function is the realtime playback service routine, so it just attempts to get a lock on the playback status data, and skips checking it if they're currently locked.